src/java/org/apache/cassandra/service/reads/ReadCallback.java

```diff
@@ -99,7 +99,13 @@ public boolean await(long timePastStart, TimeUnit unit)
     public void awaitResults() throws ReadFailureException, ReadTimeoutException
     {
         boolean signaled = await(command.getTimeout(MILLISECONDS), TimeUnit.MILLISECONDS);
-        boolean failed = failures > 0 && blockFor + failures > replicaPlan().contacts().size();
+        // Here we are checking isDataPresent in addition to the responses size because there is a possibility
+        // that an asynchronous speculative execution request could be returning after a local failure already
+        // signaled. Responses may have been set while the data reference is not yet.
+        // https://github.com/apache/cassandra/blob/ecbb970aec0780aa97406f80ea05a24b1dfa650f/src/java/org/apache/cassandra/service/reads/DigestResolver.java#L56-L59
```

Comment:
nit: If we use the JavaDoc format, you can put {@link DigestResolver#preprocess(Message)} in there and it will link in the codebase itself. (Just in case the line numbers change, etc.)